File: cpp/95. Unique Binary Search Trees II.cpp
/*95. Unique Binary Search Trees II

Given an integer n, generate all structurally unique BST's (binary search trees) that store values 1 ... n.

Example:
Input: 3
Output:
[
  [1,null,3,2],
  [3,2,null,1],
  [3,1,null,null,2],
  [2,1,3],
  [1,null,2,null,3]
]
Explanation:
The above output corresponds to the 5 unique BST's shown below:

   1         3     3      2      1
    \       /     /      / \      \
     3     2     1      1   3      2
    /     /       \                 \
   2     1         2                 3

*/

/*这道题是96.Unique Binary Search Trees的延伸，之前那个只要求算出所有不同的二叉搜索树的个数，这道题让把那些
二叉树都建立出来。这种建树问题一般来说都是用递归来解，这道题也不例外，划分左右子树，递归构造。*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<TreeNode *> generateTrees(int n) {
        if (n == 0) return {};
        return dfs(1, n);
    }
    vector<TreeNode*> dfs(int start, int end) {
        vector<TreeNode*> subTree;
        if (start > end) {
            subTree.push_back(NULL);
        }else {
            for (int i = start; i <= end; i++) {
                vector<TreeNode*> leftSubTree = dfs(start, i-1);
                vector<TreeNode*> rightSubTree = dfs(i+1, end);
                for (int j = 0; j < leftSubTree.size(); j++) {
                    for (int k = 0; k < rightSubTree.size(); k++) {
                        TreeNode *node = new TreeNode(i);
                        node->left = leftSubTree[j];
                        node->right = rightSubTree[k];
                        subTree.push_back(node);
                    }
                }
            }
        }
        return subTree;
    }
};

/*第一个版本有个严重的问题，上面的function存在大量的对象拷贝，因为所有变量都是在栈上开辟，
所以返回值的时候都需要通过拷贝构造函数来重构vector，面试中这个疏忽是不应该的。
修改版，这里应该用指针及堆来存储变量。若不用指针，全部实例化的话会存在大量的对象拷贝，要调用拷贝构造函数*/

class Solution {
public:
    vector<TreeNode *> generateTrees(int n) {
        if (n == 0) return {};
        return *dfs(1, n);
    }
    vector<TreeNode*>* dfs(int start, int end) {
        vector<TreeNode*> *subTree = new vector<TreeNode*>();
        if (start > end) {
            subTree->push_back(NULL);
        } else {
            for (int i = start; i <= end; i++) {
                vector<TreeNode*> *leftSubTree = dfs(start, i - 1);
                vector<TreeNode*> *rightSubTree = dfs(i + 1, end);
                for (int j = 0; j < leftSubTree->size(); ++j) {
                    for (int k = 0; k < rightSubTree->size(); ++k) {
                        TreeNode *node = new TreeNode(i);
                        node->left = (*leftSubTree)[j];
                        node->right = (*rightSubTree)[k];
                        subTree->push_back(node);
                    }
                }
            }
        }
        return subTree;
    }
};

File: python/95. Unique Binary Search Trees II.py
"""
95. Unique Binary Search Trees II
Medium


Share
Given an integer n, return all the structurally unique BST's (binary search trees), which has exactly n nodes of unique values from 1 to n. Return the answer in any order.

 

Example 1:
Input: n = 3
Output: [[1,null,2,null,3],[1,null,3,2],[2,1,3],[3,1,null,null,2],[3,2,null,1]]

Example 2:
Input: n = 1
Output: [[1]]
"""


"""
Define functions fn(low, high) to return the trees using numbers between low (inclusive) 
and hi (exclusive). In this way, the trees can be built recursively. For example, at node i, 
its left subtree can be built from fn(lo, i) and its right subtree can be built from fn(i+1, hi).

"""


"""
[Python] DFS with Memoization - Clean & Concise
107
hiepit's avatar
hiepit

34438
Last Edit: September 2, 2021 2:34 AM

4.3K VIEWS

Idea

Let dfs(left, right) return all valid BSTs where values in the BST in range [left..right].
Then dfs(1, n) is our result.
To solve dfs(left, right), we just
Generate root value in range [left...right]
Get left subtrees by leftNodes = dfs(left, root-1)
Get right subtrees by rightNodes = dfs(root+1, right).
Add all combination between leftNodes and rightNodes to form root trees.
Can we cache the result of dfs(left, right) to prevent it to re-compute multiple time.
There is a simillar problem, which is 894. All Possible Full Binary Trees, try to solve it yourself.

Complexity

Time: O(C0+C1+...Cn), where Cn is the Catalan number, n <= 8. Can check this problem 96. Unique Binary Search Trees to know why the number of nodes in the BST with n nodes is a Catalan number.
The Catalan numbers for n = 0, 1, 2, 3, 4, 5, 6, 7, 8 are 1, 1, 2, 5, 14, 42, 132, 429, 1430.
Space: O(n * Cn), there is total Cn BSTs, each BST has n nodes.
"""


class Solution:
    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:

        @lru_cache(None)
        def dfs(left, right):

            if left > right:
                return [None]

            if left == right:
                return [TreeNode(left)]

            res = []
            for rval in range(left, right+1):
                leftNodes = dfs(left, rval - 1)
                rightNodes = dfs(rval+1, right)

                for left in leftNodes:
                    for right in rightNodes:
                        node = TreeNode(rval, left, right)
                        res.append(node)
            return res

        return dfs(1, n)


class Solution:
    def generateTrees(self, n: int) -> List[TreeNode]:
        if not n:
            return []

        @lru_cache(None)
        def dfs(low, high):
            """Return root of tree using nums from low (inclusive) to high (exclusive)"""
            if low == high:
                return [None]

            res = []
            for node in range(low, high):
                for left in dfs(low, node):
                    for right in dfs(node+1, high):
                        res.append(TreeNode(node, left, right))

            return res

        return dfs(1, n+1)
